File: day01/main.go
package main

import (
	"io"
	"sort"
	"strings"

	"github.com/strangelytyped/advent-of-code-2022/utils"
)

func parseElves(input string) []int {
	sum := 0
	result := []int{}

	for _, line := range strings.Split(input, "\n") {
		if len(strings.TrimSpace(line)) == 0 {
			result = append(result, sum)
			sum = 0
		} else {
			sum += utils.AtoiOrPanic(line)
		}
	}
	if sum != 0 {
		result = append(result, sum)
	}

	return result
}

func Part1(input io.Reader) int {
	maxElf := 0
	for _, elf := range parseElves(utils.ReadInput(input)) {
		if elf > maxElf {
			maxElf = elf
		}
	}
	return maxElf
}

func Part2(input io.Reader) int {
	elves := parseElves(utils.ReadInput(input))
	sort.Ints(elves)
	count := len(elves)
	return elves[count-1] + elves[count-2] + elves[count-3]
}

func main() {
	utils.Run("day01.txt", Part1, Part2)
}

File: day01/main.mjs
import * as utils from '../utils/utils.mjs';

async function* parseElves(readStream){
    let elf = 0;
    for await (const line of readStream) {
        if(!line.trim().length){
            yield elf;
            elf = 0;
        } else {
            elf += parseInt(line)
        }
    }
    yield elf;
}

function sortedInsert(arr, n){
  let i = 0;
  while(arr[i] < n){
    i++;
  }
  arr.splice(i, 0, n)
  return arr;
}

async function solve(readStream, n){
    let vals = [];

    for await (const elf of parseElves(readStream)){
        if (vals.length < n){
            vals = sortedInsert(vals, elf)
        } else if (elf > vals[0]){
            vals = sortedInsert(vals.slice(1), elf)
        }
        
    }
    return vals.reduce((accum, i) => accum + i, 0)
}


async function part1(readStream){
    return solve(readStream, 1);
}

async function part2(readStream){
    return solve(readStream, 3);
}



const testInput = `
1000
2000
3000

4000

5000
6000

7000
8000
9000

10000
`

utils.run(testInput, 24000, 45000, "../inputs/day01.txt", part1, part2)
